### DISCUSSION TITLE
JoinHandle.abort

### DISCUSSION THREAD

**Discussion by itamarsch** [tokio-rs/tokio#5534] [Q&A] [answered]
What does this function actually do?

What does it do the memory the closure has ownership of?

Is it a bad practice?

If it is a bad practice what is a better way to stop a running rask?

Does it cause UB?

**Comment by Darksonn** [comment, marked as answer]
The `abort` method will cancel the task next time it yields at an `.await`. For tasks that don't [block the thread](https://ryhl.io/blog/async-what-is-blocking/), this means that it will be cancelled immediately. 

Cancellation is done by running the destructor of the future, which runs the destructor of all local variables defined inside the future. This also cleans up all memory owned by the future. 

Using `abort` is not bad practice — it works great as long as you don't need to do any asynchronous cleanup. For cases where `abort` is not sufficient, you can use the techniques outlined in [the page on graceful shutdown](https://tokio.rs/tokio/topics/shutdown) on the Tokio website.

It certainly does not cause UB. Like the standard library, Tokio promises that causing UB with only safe code is impossible. Since you can call `abort` without unsafe, Tokio promises that it must never cause UB.

**Reply by itamarsch** [reply]
Cool thank you very much!

**Reply by ShaneMurphy2** [reply]
I feel like the first sentence of your answer should be in the docs for `abort`. I was very curious about the actual behavior, even though cancelling at the next await point is logically the only thing Tokio could really do.

**Reply by Darksonn** [reply]
@ShaneMurphy2 Would you be willing to open a PR to add it? Or open a documentation bug saying that it should be added?

**Reply by ShaneMurphy2** [reply]
Looks like a section for this was already [added on master](https://github.com/tokio-rs/tokio/blob/master/tokio/src/task/mod.rs#L115C1-L115C22).

**Comment by carllerche** [comment]
@itamarsch If the docs are lacking, we would appreciate a PR to improve them. You can try to expand on them using what you learned here.

**Comment by ShaneMurphy2** [comment]
Sure!
